keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -0,0 +1,483 @@
+        // the endpoint serving the resource.
+        // Immutable.
+        Resource GroupVersionResource `json:"resource"`
+        // The token used in the list options to get the next chunk of objects
+        // to migrate. When the .status.conditions indicates the migration is
+        // "Running", users can use this token to check the progress of the
+        // migration.
+        // +optional
+        ContinueToken string `json:"continueToken,omitempty"`
+        // The storage version hash to avoid races.
+        StorageVersionHash string `json:"storageVersionHash,omitempty"`
+    }
+
+    // The names of the group, the version, and the resource.
+    type GroupVersionResource struct {
+        // The name of the group.
+        Group string `json:"group,omitempty"`
+        // The name of the version.
+        Version string `json:"version,omitempty"`
+        // The name of the resource.
+        Resource string `json:"resource,omitempty"`
+    }
+    
+    // Status of the storage version migration.
+    type StorageVersionMigrationStatus struct {
+        // The latest available observations of the migration's current state.
+        // +optional
+        // +patchMergeKey=type
+        // +patchStrategy=merge
+        Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+    }
+
+    // StorageVersionMigrationList is a collection of storage version migrations.
+    type StorageVersionMigrationList struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ListMeta `json:"metadata,omitempty"`
+        // Items is the list of StorageVersionMigration
+        Items []StorageVersionMigration `json:"items"`
+    }
+    ```
+
+- APIs in-tree will be _converted to `built-in types`_ from CRD.
+
+#### [Undecided] Changes while we move above APIs in-tree:
+To avoid any conflicts with the Storage Version Migrators running out of tree, we will change the _`group`_ from `migration.k8s.io` to `storagemigration.k8s.io`.
+
+The final APIs that will be moved in-tree are:
+- `v1alpha1` of `storageversionmigrations.storagemigration.k8s.io`
+
+### [Controller](https://github.com/kubernetes-sigs/kube-storage-version-migrator/tree/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/controller) to move
+#### [Migrator Controller](https://github.com/kubernetes-sigs/kube-storage-version-migrator/tree/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/migrator)
+Currently, the Storage Version Migrator comprises two controllers: the `Trigger` controller and the `Migrator` controller. The Trigger controller performs resource discovery, identifying supported resources with the preferred server version every `10 minutes`. Subsequently, the Trigger controller creates the `StorageVersionMigration` resource to initiate the migration process. The Migrator controller then picks up this resource and executes the actual migration.
```

Comment:
Is this design level or edge triggered? If edge triggered, how do we ensure that we don't miss (fail to observe) storage version changes?

Reply:
Okay, I think this answers my question.  Can we make it super clear in the KEP that there will be no automatic triggering, I somehow missed that when first reading it.

Reply:
In `Proposal` section, I updated this:
```
**[UNCLEAR]** Create a new controller that watches the storage version API and automatically triggers migrations based on changes (to the SV API)
```

to:

```
For Alpha, we will not trigger automatic storage version migration, and it will be deferred to the user.
```